dwave/preprocessing/include/dwave/presolve.h

```diff
@@ -0,0 +1,410 @@
+        : model_(std::move(model)), postsolver_(), default_techniques_(), detached_(false) {}
+
+template <class bias_type, class index_type, class assignment_type>
+void Presolver<bias_type, index_type, assignment_type>::apply() {
+    if (detached_) throw std::logic_error("model has been detached, presolver is no longer valid");
+    // do nothing if they haven't been loaded
+    // todo: actually read from a vector of techniques or similar
+    if (!default_techniques_) return;
+
+    // One time techniques ----------------------------------------------------
+
+    // *-- spin-to-binary
+    for (size_type v = 0; v < model_.num_variables(); ++v) {
+        if (model_.vartype(v) == dimod::Vartype::SPIN) {
+            postsolver_.substitute_variable(v, 2, -1);
+            model_.change_vartype(dimod::Vartype::BINARY, v);
+        }
+    }
+
+    // *-- remove offsets
+    for (size_type c = 0; c < model_.num_constraints(); ++c) {
+        auto& constraint = model_.constraint_ref(c);
+        if (constraint.offset()) {
+            constraint.set_rhs(constraint.rhs() - constraint.offset());
+            constraint.set_offset(0);
+        }
+    }
+
+    // *-- flip >= constraints
+    for (size_type c = 0; c < model_.num_constraints(); ++c) {
+        auto& constraint = model_.constraint_ref(c);
+        if (constraint.sense() == dimod::Sense::GE) {
+            constraint.scale(-1);
+        }
+    }
+
+    // *-- remove self-loops
+    substitute_self_loops();
+
+    // Trivial techniques -----------------------------------------------------
+
+    bool changes = true;
+    const index_type max_num_rounds = 100;  // todo: make configurable
+    for (index_type num_rounds = 0; num_rounds < max_num_rounds; ++num_rounds) {
+        if (!changes) break;
+        changes = false;
+
+        // *-- todo: clear out 0 variables/interactions in the constraints
+
+        // *-- todo: check for NAN
+
+        // *-- remove single variable constraints
+        size_type c = 0;
+        while (c < model_.num_constraints()) {
+            auto& constraint = model_.constraint_ref(c);
+
+            if (constraint.num_variables() == 0) {
+                // remove after checking feasibity
+                if (constraint.offset() != constraint.rhs()) {
+                    throw std::logic_error("infeasible");
+                }
+                model_.remove_constraint(c);
+                changes = true;
+                continue;
+            } else if (constraint.num_variables() == 1 && !constraint.is_soft()) {
+                index_type v = constraint.variables()[0];
+
+                // ax ◯ c
+                bias_type a = constraint.linear(v);
+                assert(a);  // should have already been removed if 0
+
+                // offset should have already been removed but may as well be safe
+                bias_type rhs = (constraint.rhs() - constraint.offset()) / a;
```

Comment:
I think this could be +-inf which could result in bad bounds

Reply:
You're right, we should test for inf, and nan as well. Will do that in a followup at some point